### PR TITLE
Implement authentication #JC19-457

### PR DESCRIPTION
**Pull request by aletor123**

### CODE REVIEW

File: apps/react/src/utils/apiUrls.ts [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+  /** Get authentication API urls. */
+  public static get authUrls(): AuthUrls {
+    const authPrefix = 'auth/';
+    return {
+      login: `${authPrefix}login/`,
+      register: `${authPrefix}register/`,
+      tokenRefresh: `${authPrefix}token/refresh/`,
+    };
+  }
+}
```
Comment (timofei-maximov-saritasa) [inline apps/react/src/utils/apiUrls.ts:26]: Do you really need this as class?

File: libs/core/mappers/appError.mapper.ts [hunk 2]
```diff
@@ -0,0 +1,48 @@
+import { AxiosError } from 'axios';
+
+import { ValidationErrorDto } from '../dtos/validationError.dto';
+import { AppError, AppValidationError } from '../models/appError';
+
+import { ValidationErrorMapper } from './validation-mappers/validationError.mapper';
+
+/** Mapper for app errors. */
+export class AppErrorMapper {
```
Comment (timofei-maximov-saritasa) [inline libs/core/mappers/appError.mapper.ts:9]: Please, be consistent with your mappers. You have `namespace` and `class` mappers. Select one approach.

**Review by timofei-maximov-saritasa — CHANGES_REQUESTED** [review]

File: apps/react/src/features/auth/pages/LoginPage.tsx [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+    }
+  }, [formik.values]);
+
+  /** Clear login error on unmount. */
+  useEffect(() => () => {
+    dispatch(clearLoginError());
+  }, []);
+
+  return (
+    <div className='form-wrapper'>
```
[the remark this answers is not in the document]
Comment (aletor123) [reply apps/react/src/features/auth/pages/LoginPage.tsx:65]: What if these are global styles?

File: apps/react/src/components/Navbar/NavBar.tsx [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+      {!isUserLoading && (
+        <>
+          {user === null &&
+        <Button component={Link} color="inherit" to={routePaths.login}>Login</Button>
+          }
+          {user !== null && (
+            <>
+              <Typography>{user.email}</Typography>
+              <Button color="inherit" onClick={onLogoutClick}>Logout</Button>
+            </>
```
Comment (timofei-maximov-saritasa) [inline apps/react/src/components/Navbar/NavBar.tsx:43]: You can you ternary operator here https://beta.reactjs.org/learn/conditional-rendering#conditional-ternary-operator--

File: apps/react/src/features/auth/components/RegisterForm.tsx [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+        {...formik}
+      />
+      <TextField
+        name="passwordConfirmation"
+        autoComplete="current-password"
+        label="Confirm password"
+        type="password"
+        {...formik}
+      />
+      {formError !== null && <p className="text-danger">{formError}</p>}
```
Comment (timofei-maximov-saritasa) [inline apps/react/src/features/auth/components/RegisterForm.tsx:112]: You can use `Typography` component with `sx` property. `sx` to keep theme consistent across your app.
Also, you can use `Alert` component with built-in `error` variant

```tsx
<Typography
  sx={theme => ({
    color: theme.palette.error.main,
  })}
>
 ```
 
 Or you can use

File: apps/react/src/features/auth/components/RegisterForm.tsx [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+      <div className="form__controls">
+        <Button type="submit" variant="contained" disabled={!formik.isValid}>
+            Submit
+        </Button>
```
Comment (timofei-maximov-saritasa) [inline apps/react/src/features/auth/components/RegisterForm.tsx:116]: Formatting issue

File: apps/react/src/features/auth/pages/RegisterPage.tsx [hunk 7]
```diff
@@ -0,0 +1,12 @@
+import { FC, memo } from 'react';
+
+import { RegisterForm } from '../components/RegisterForm';
+
+const RegisterPageComponent: FC = () => (
+  <div className="form-wrapper">
```
Comment (timofei-maximov-saritasa) [inline apps/react/src/features/auth/pages/RegisterPage.tsx:6]: I'm not sure that you should use global css for this. Actually, you can make standalone component. like `FormLayout` and use css modules  to get reduce duplicated code.

File: apps/react/src/features/auth/components/LoginForm.tsx [hunk 8]
```diff
@@ -0,0 +1,93 @@
+import { FC, memo, useEffect, useState } from 'react';
```
Comment (timofei-maximov-saritasa) [inline apps/react/src/features/auth/components/LoginForm.tsx:1]: Put this component in folder with the same name as component's file name

File: apps/react/src/features/auth/pages/LoginPage.tsx [hunk 9]
```diff
@@ -0,0 +1,12 @@
+import { FC, memo } from 'react';
```
Comment (timofei-maximov-saritasa) [inline apps/react/src/features/auth/pages/LoginPage.tsx:1]: Same about folder

File: apps/react/src/components/PageLoader/PageLoader.tsx [hunk 10]
```diff
@@ -0,0 +1,16 @@
+import Backdrop from '@mui/material/Backdrop';
```
Comment (timofei-maximov-saritasa) [inline apps/react/src/components/PageLoader/PageLoader.tsx:1]: Don't forget to add `index.ts` file for re-exporting your components. Let's be consistent 😇

File: apps/react/src/components/form/Form/Form.tsx [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+
+import styles from './Form.module.css';
+
+interface Props extends FormHTMLAttributes<HTMLFormElement> {
+
+  /** Child node. */
+  readonly children: ReactNode;
+}
+
+const FormComponent: FC<Props> = ({ children, ...props }: Props) => (
```
Comment (timofei-maximov-saritasa) [inline apps/react/src/components/form/Form/Form.tsx:11]:
```suggestion
const FormComponent: FC<Props> = ({ children, ...props }) => (
```

File: apps/react/src/components/form/FormControls/FormControls.tsx [hunk 12]
```diff
@@ -0,0 +1,17 @@
...
+
+import styles from './FormControls.module.css';
+
+interface Props extends HTMLAttributes<HTMLDivElement> {
+
+  /** Child node. */
+  readonly children: ReactNode;
+}
+
+const FormControlsComponent: FC<Props> = ({ children, ...props }: Props) => (
```
Comment (timofei-maximov-saritasa) [inline apps/react/src/components/form/FormControls/FormControls.tsx:11]:
```suggestion
const FormControlsComponent: FC<Props> = ({ children, ...props }) => (
```

File: apps/react/src/components/form/FormWrapper/FormWrapper.tsx [hunk 13]
```diff
@@ -0,0 +1,17 @@
...
+
+import styles from './FormWrapper.module.css';
+
+interface Props extends HTMLAttributes<HTMLDivElement> {
+
+  /** Child node. */
+  readonly children: ReactNode;
+}
+
+const FormWrapperComponent: FC<Props> = ({ children, ...props }: Props) => (
```
Comment (timofei-maximov-saritasa) [inline apps/react/src/components/form/FormWrapper/FormWrapper.tsx:11]:
```suggestion
const FormWrapperComponent: FC<Props> = ({ children, ...props }) => (
```

**Review by timofei-maximov-saritasa — APPROVED** [review]

File: apps/react/src/features/auth/components/RegisterForm/RegisterForm.tsx [hunk 14]
```diff
@@ -0,0 +1,141 @@
...
+import { clearRegistrationError } from '@js-camp/react/store/auth/slice';
+
+import { TextField } from 'formik-mui';
+
+import { routePaths } from '../../../../utils/routePaths';
+import { passwordYupValidator } from '../../../../utils/forms';
+import { Form } from '../../../../components/form/Form';
+import { FormControls } from '../../../../components/form/FormControls';
+
+const RegisterValidationSchema = Yup.object().shape({
```
Comment (golovachev-yaroslav) [inline apps/react/src/features/auth/components/RegisterForm/RegisterForm.tsx:20]: Let's move initial state and validation schema to the separate file(ex.: formSettings.js). It makes this code more clear and without overload.

**Review by golovachev-yaroslav — CHANGES_REQUESTED** [review]

**Review by pavelzadorin-saritasa — APPROVED** [review]

**Pull request merged**